### PR TITLE
Feature/upgradable_contract

### PR DESCRIPTION
**Pull request by softskillpro**
Upgradable contract

- [ ] https://github.com/plexus-money/contracts/issues/14

### CODE REVIEW

**Comment by terminator0x** [comment]
@softskillpro is this PR ready for review?

**Comment by softskillpro** [comment]
> @softskillpro is this PR ready for review?
Yes. I fixed conflicts and pushed it. it works on my local properly.   
But currently I see following issues with github action. 
"An unexpected error occurred:
ProviderError: Must be authenticated!"
it looks like hadhat config issue - RPC_NODE_URL

**Comment by terminator0x** [comment]
> > @softskillpro is this PR ready for review?
> 
> Yes. I fixed conflicts and pushed. it works on my local properly.
> But currently, I see the following issue with GitHub action.
> "An unexpected error occurred: ProviderError: Must be authenticated!"
> It looks like hardhat config issue - RPC_NODE_URL

Pleas rebase against the master branch, so that you get the updated github actions file and the tests will start working.

**Review by terminator0x — CHANGES_REQUESTED** [review]
@softskillpro Great work, so far, I love it 😃 . But still their are a few things that need to be sorted.

1.  Please import the `Ownable` contracts from the `@openzeppelin` node.js lib 
   i.e. for the `Ownable.sol` file you do the following

```sol
import "@openzeppelin/contracts/access/Ownable.sol";
```
- The reason for this is because openzeppelin contracts are trusted and hence it makes auditing our contracts a much easier task.

2. Add more tests to verify that the ownership of the deployed contracts belongs to the respective proxy with each new upgrade, this is crucial because a compromise would upet the whole plexus eco-system.
3. Add a test to check that the ownership of the proxy contracts remains to be the original deployer account and this shouldn't change at all regardless of how many upgrades are done.
4. Finally add another test to verify that the global state variables for the `tokenrewards.sol` contract will always remain the same regardless of how many times the contract is deployed.
i.e. the `stakingTokenWhitelist`, `stakingTokensAddress`, `stakingLPTokensAddress` and the rest of the global public vars should remain the same in the proxy contract after 3 deployments of the same contract in different addresses
3.

**Comment by softskillpro** [comment]
> @softskillpro Great work, so far, I love it 😃 . But still their are a few things that need to be sorted.
> 
> 1. Please import the `Ownable` contracts from the `@openzeppelin` node.js lib
>    i.e. for the `Ownable.sol` file you do the following
> 
> ```gerber
> import "@openzeppelin/contracts/access/Ownable.sol";
> ```
> 
> * The reason for this is because openzeppelin contracts are trusted and hence it makes auditing our contracts a much easier task.
> 
> 1. Add more tests to verify that the ownership of the deployed contracts belongs to the respective proxy with each new upgrade, this is crucial because a compromise would lead to loss of funds for the whole plexus eco-system.
> 2. Add a test to check that the ownership of the proxy contracts remains to be the original deployer account and this shouldn't change at all regardless of how many upgrades are done.
> 3. Finally add another test to verify that the global state variables for the `tokenrewards.sol` contract will always remain the same regardless of how many times the contract is deployed.
>    i.e. the `stakingTokenWhitelist`, `stakingTokensAddress`, `stakingLPTokensAddress` and the rest of the global public vars should remain the same in the proxy contract after 3 deployments of the same contract in different addresses

**Comment by softskillpro** [comment]
@terminator0x 
1. Please import the Ownable contracts from the @openzeppelin node.js lib
import "@openzeppelin/contracts/access/Ownable.sol";

Ownable.sol of openzeppelin doesn't have functions for updating owner, as our contracts need to this feature.
Most of the contracts has changeOwner() functions.
So we need to customize @openzeppelin/contracts/access/Ownable.sol for using in our contracts.
I'll copy the source and add a new function that update owner, Is it okay?
for example, function updateOwner(address newOwner) public onlyOwner {}

**Comment by softskillpro** [comment]
@terminator0x Ownable.sol of openzeppelin doesn't have functions for updating owner, as our contracts need to this feature.
Most of the contracts has changeOwner() functions.
So we need to customize @openzeppelin/contracts/access/Ownable.sol for using in our contracts.
I'll copy the source and add a new function that update owner, Is it okay?
for example, function updateOwner(address newOwner) public onlyOwner {}

**Comment by softskillpro** [comment]
> > @softskillpro Great work, so far, I love it 😃 . But still their are a few things that need to be sorted.
> > 
> > 1. Please import the `Ownable` contracts from the `@openzeppelin` node.js lib
> >    i.e. for the `Ownable.sol` file you do the following
> > 
> > ```
> > import "@openzeppelin/contracts/access/Ownable.sol";
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * The reason for this is because openzeppelin contracts are trusted and hence it makes auditing our contracts a much easier task.
> > 
> > 
> > 1. Add more tests to verify that the ownership of the deployed contracts belongs to the respective proxy with each new upgrade, this is crucial because a compromise would lead to loss of funds for the whole plexus eco-system.
> > 2. Add a test to check that the ownership of the proxy contracts remains to be the original deployer account and this shouldn't change at all regardless of how many upgrades are done.
> > 3. Finally add another test to verify that the global state variables for the `tokenrewards.sol` contract will always remain the same regardless of how many times the contract is deployed.
> >    i.e. the `stakingTokenWhitelist`, `stakingTokensAddress`, `stakingLPTokensAddress` and the rest of the global public vars should remain the same in the proxy contract after 3 deployments of the same contract in different addresses

@terminator0x Ownable.sol of openzeppelin doesn't have functions for updating owner, as our contracts need to this feature.
Most of the contracts has changeOwner() functions.
So we need to customize @openzeppelin/contracts/access/Ownable.sol for using in our contracts.
I'll copy the source and add a new function that update owner, Is it okay?
for example, function updateOwner(address newOwner) public onlyOwner {}

File: test/8_proxy.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        expect(prevWrapperProxyOwner).to.equal(prevWrapperContractOwner);
+        expect(prevWrapperSushiProxyOwner).to.equal(prevWrapperSushiContractOwner);
+        expect(prevTokenRewardsProxyOwner).to.equal(prevTokenRewardsContractOwner);
+        expect(prevPlexusOracleProxyOwner).to.equal(prevPlexusOracleContractOwner);
+        expect(prevTier1StakingProxyOwner).to.equal(prevTier1StakingContractOwner);
+        expect(prevCoreProxyOwner).to.equal(prevCoreContractOwner);
+      });
+    });
+
+    describe('', () => {
```
Comment (terminator0x) [inline test/8_proxy.js:96]: Remove the describe section, it's unnecessary

File: test/8_proxy.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    const prevTier2FarmContract = await ethers.getContractAt("Tier2FarmController", await tier2Farm.target());
+    const prevTier2FarmContractOwner = await prevTier2FarmContract.owner();
+    const prevTier2AaveProxyOwner = await tier2Aave.owner();
+    const prevTier2AaveContract = await ethers.getContractAt("Tier2AaveFarmController", await tier2Aave.target());
+    const prevTier2AaveContractOwner = await prevTier2AaveContract.owner();
+    const prevTier2PickleProxyOwner = await tier2Pickle.owner();
+    const prevTier2PickleContract = await ethers.getContractAt("Tier2PickleFarmController", await tier2Pickle.target());
+    const prevTier2PickleContractOwner = await prevTier2PickleContract.owner();
+
+    describe('', () => {
```
Comment (terminator0x) [inline test/8_proxy.js:85]: Remove the describe function here

File: test/8_proxy.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    const newWrapperSushiContract = await ethers.getContractAt("WrapAndUnWrapSushi", await wrapperSushi.target());
+    const newTokenRewardsContract = await ethers.getContractAt("TokenRewards", await tokenRewards.target());
+    const newPlexusOracleContract = await ethers.getContractAt("PlexusOracle", await plexusOracle.target());
+    const newTier1StakingContract = await ethers.getContractAt("Tier1FarmController", await tier1Staking.target());
+    const newCoreContract = await ethers.getContractAt("Core", await core.target());
+    const newTier2FarmContract = await ethers.getContractAt("Tier2FarmController", await tier2Farm.target());
+    const newTier2AaveContract = await ethers.getContractAt("Tier2AaveFarmController", await tier2Aave.target());
+    const newTier2PickleContract = await ethers.getContractAt("Tier2PickleFarmController", await tier2Pickle.target());
+
+    describe('', () => {
```
Comment (terminator0x) [inline test/8_proxy.js:159]: Same here remove the empty describe section, this should only be used to literally describe a testing section, so we can't have it blank

File: test/8_proxy.js [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        expect(prevWrapperContractOwner).to.equal(await newWrapperContract.owner());
+        expect(prevWrapperSushiProxyOwner).to.equal(await newWrapperSushiContract.owner());
+        expect(prevTokenRewardsProxyOwner).to.equal(await newTokenRewardsContract.owner());
+        expect(prevPlexusOracleProxyOwner).to.equal(await newPlexusOracleContract.owner());
+        expect(prevTier1StakingProxyOwner).to.equal(await newTier1StakingContract.owner());
+        expect(prevCoreProxyOwner).to.equal(await newCoreContract.owner());
+      });
+    });
+
+    describe('', () => {
```
Comment (terminator0x) [inline test/8_proxy.js:170]: Same here remove empty describe section and leave the it function

File: test/8_proxy.js [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+      it('For tier2Farm, tier2Aave, and tier2Pickle, The Owner of proxy should be not equal to the one of his updated logic contract', async () => {
+        expect(prevTier2FarmProxyOwner).to.not.equal(await newTier2FarmContract.owner());
+        expect(prevTier2AaveProxyOwner).to.not.equal(await newTier2AaveContract.owner());
+        expect(prevTier2PickleProxyOwner).to.not.equal(await newTier2PickleContract.owner());
+      });
+    })
+
+    describe('', () => {
```
Comment (terminator0x) [inline test/8_proxy.js:178]: Same here remove empty describe section and leave the it function

File: test/8_proxy.js [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+        expect(await plexusOracle.target()).to.not.equal(prevPlexusOracleTarget);
+        expect(await tier1Staking.target()).to.not.equal(prevTier1StakingTarget);
+        expect(await core.target()).to.not.equal(prevCoreTarget);
+        expect(await tier2Farm.target()).to.not.equal(prevTier2FarmTarget);
+        expect(await tier2Aave.target()).to.not.equal(prevTier2AaveTarget);
+        expect(await tier2Pickle.target()).to.not.equal(prevTier2PickleTarget);
+      });
+    });
+
+    describe('', () => {
```
Comment (terminator0x) [inline test/8_proxy.js:192]: Same here remove empty describe section and leave the it function

File: contracts/proxyLib/Context.sol [hunk 7]
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: MIT
```
Comment (terminator0x) [inline contracts/proxyLib/Context.sol:1]: Import this file from the openzeppelin node.js lib

**Review by terminator0x — CHANGES_REQUESTED** [review]

**Comment by terminator0x** [comment]
Rebased into #29

**Pull request closed without merging**